scipy/sparse/data.py

```diff
@@ -64,8 +64,16 @@ def __itruediv__(self, other):  # self /= other
         else:
             return NotImplemented
 
-    def astype(self, t):
-        return self._with_data(self._deduped_data().astype(t))
+    def astype(self, dtype, casting='unsafe', copy=True):
+        dtype = np.dtype(dtype)
+        if self.dtype != dtype:
+            return self._with_data(
+                self._deduped_data().astype(dtype, casting=casting, copy=copy),
+                copy=False)
```

Comment:
I think `copy=copy` is still right because it's okay to share indices for `.astype(..., copy=False)`, even if the data arrays are different. I see the `copy` argument as meaning:
 * `copy=True`: make sure the result shares nothing with `self`
 * `copy=False`: allow the result to share memory with `self`, but don't require it

An actual use-case might look like `foo = foo.astype(float, copy=False)`, where I don't care about the previous value of `foo` at all and want to avoid unneeded copies where possible.